[RNMobile] Fetch VideoPress metadata for native block

SiobhyB opened the pull request:
Part of a series of PRs to address https://github.com/wordpress-mobile/gutenberg-mobile/issues/5611. 

## Related PRs

The changes in this PR relies on the following:

* `gutenberg-mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/5652

## Proposed changes:

* https://github.com/Automattic/jetpack/pull/29997/commits/06b3340d5223c4a23abb8a3a936846445178ffbe: Skip an unnecessary `isUserConnected` check. As only simple WordPress.com sites are supported on native for the time being, we can safely assume that users will be connected to Jetpack. We can revisit this in the future when expanding support. 
* https://github.com/Automattic/jetpack/pull/29997/commits/fe1ba4d1dbb4844d0ba6f90f97cb5c10ba9db198: To avoid creating native-specific files with very few changes, the `debug` functionality (which isn't compatible with native by default) is conditionally changed to `console.log`. 
* https://github.com/Automattic/jetpack/pull/29997/commits/aa4bf9b3244f3a7af9f72fc641be29732d008347: We pass the endpoint as a `path` on native, to ensure the URL is correctly built in the native apps and to allows us more control over supported endpoints. Support for this endpoint is added in https://github.com/wordpress-mobile/gutenberg-mobile/pull/5652.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

p9ugOq-3mV-p2 

## Does this pull request change what data or activity we track or use?

No, it does not.

## Testing instructions:

> **Note**
> The changes in this PR make sure the VideoPress metadata is synced from the web → app only. The necessary changes to ensure the metadata is also synced from the app → web will come in a future PR.

* As the `useSyncMedia` hook has not yet been implemented on native, it's necessary to first apply the following patch before testing these changes.

<details>
 <summary>Patch here ⤵️</summary>

These changes will implement the `useSyncMedia` hook and log the results of `useVideoData`: 

```patch
diff --git a/projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx b/projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx
index ff9906b1bf..ec61816006 100644
--- a/projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx
+++ b/projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx
@@ -13,6 +13,7 @@ import { useDispatch, useSelect } from '@wordpress/data';
 import { useState, useCallback, useEffect } from '@wordpress/element';
 import { __ } from '@wordpress/i18n';
 import { store as noticesStore } from '@wordpress/notices';
+import { useSyncMedia } from '../../hooks/use-sync-media';
 /**
  * External dependencies
  */
@@ -133,6 +134,8 @@ export default function VideoPressEdit( {
 		[ setAttributes ]
 	);
 
+	useSyncMedia( attributes, setAttributes );
+
 	const handleDoneUpload = useCallback(
 		newVideoData => {
 			setIsUploadingFile( false );
diff --git a/projects/packages/videopress/src/client/block-editor/hooks/use-sync-media/index.ts b/projects/packages/videopress/src/client/block-editor/hooks/use-sync-media/index.ts
index 66ec465133..5c1a39dc9d 100644
--- a/projects/packages/videopress/src/client/block-editor/hooks/use-sync-media/index.ts
+++ b/projects/packages/videopress/src/client/block-editor/hooks/use-sync-media/index.ts
@@ -279,7 +279,7 @@ export function useSyncMedia(
 	 * Store and compare the block attributes
 	 * in order to detect changes on them.
 	 */
-	const { isGeneratingPoster } = useVideoPosterData( attributes );
+	const isGeneratingPoster = false;
 
 	/*
 	 * Block attributes => Media data (sync)
diff --git a/projects/packages/videopress/src/client/block-editor/hooks/use-video-data/index.ts b/projects/packages/videopress/src/client/block-editor/hooks/use-video-data/index.ts
index f5b3b0816c..22c29c21b7 100644
--- a/projects/packages/videopress/src/client/block-editor/hooks/use-video-data/index.ts
+++ b/projects/packages/videopress/src/client/block-editor/hooks/use-video-data/index.ts
@@ -63,6 +63,10 @@ export default function useVideoData( {
 						skipRatingControl,
 					} );
 
+					const responseString = JSON.stringify( response );
+
+					console.log( 'video data: ' + responseString );
+
 					if ( response.duration ) {
 						debug(
 							`video duration available: ${ response.duration }, retried ${ retries } times`,
```
</details>

* With the patch applied, first navigate to the post editor on the web.
* Either add a new VideoPress block or use an existing one, and make sure to tweak its settings.
* It's particularly important to tweak the settings in the **Details** and **Privacy and rating** panels,  as these are the settings that are not currently being synced.
* Save your change on the web and then navigate to the post in the native app.
* Verify that the block's settings have been correctly fetched from the web.
* Verify also that `video data: {video data}` has been correctly logged to the console.
* Go through a few tries of editing the block on the web then returning the app to ensure syncing is working from the web → app.

SiobhyB committed:
refactor: Enable debugging on native

SiobhyB committed:
refactor: Skip unnecessary isUserConnected check

SiobhyB committed:
docs: Add details of changes to CHANGELOG

SiobhyB committed:
Merge branch 'trunk' into rnmobile/methods-to-fetch-metadata

SiobhyB committed:
refactor: Replace 'isWeb' conditional w/'isNative'

SiobhyB committed:
refactor: Set up different endpoint for native

SiobhyB committed:
refactor: Replace isWeb conditionals with isNative

SiobhyB committed:
refactor: Fix isNative checks

SiobhyB commented:
@jhnstn, @fluiddot, just noting that I've updated this PR based on feedback in the (now closed) Gutenberg PR. It's ready for review again.

fluiddot commented on projects/packages/videopress/src/client/block-editor/hooks/use-video-data/index.ts:
```diff
@@ -14,7 +14,10 @@ import { fetchVideoItem } from '../../../lib/fetch-video-item';
+const isNative = Platform.isNative;
+
+// eslint-disable-next-line no-console
+const debug = isNative ? console.log : debugFactory( 'videopress:video:use-video-data' );
```
I noticed that calls to `debug` are not being displayed in the native version because it's not enabled. If we add the following line all debug logs are printed:
```
debugFactory.enable( 'videopress:*' );
```

Since `enable` function should only be called once, I'm wondering if we could add this to [the entry point](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/videopress/src/client/block-editor/editor.native.js):
```
import debugFactory from 'debug';
import registerVideoPressBlock from './blocks/video/index';

// eslint-disable-next-line no-undef
if ( __DEV__ ) {
	debugFactory.enable( 'videopress:*' );
}

registerVideoPressBlock();

```
WDYT?

SiobhyB replied:
Much nicer, thanks Carlos! Done in https://github.com/Automattic/jetpack/pull/29997/commits/08f9a92bf86dd2d3dae65afac390a9c7d6a409d1.

fluiddot reviewed (commented):
@SiobhyB On iOS works as expected, I tried modifying the settings and confirmed they are fetched properly on the native version. However, on Android using an actual device, I'm getting the following error:
```
Possible Unhandled Promise Rejection (id: 0):
Error: undefined is not a function
Error: undefined is not a function
```

I tried to debug it via the Chrome debugger and it doesn't happen. This makes me think that there's a call somewhere within the `useVideoData` hook that is not supported in Hermes.

SiobhyB committed:
Merge branch 'trunk' into rnmobile/methods-to-fetch-metadata

SiobhyB committed:
refactor: Replace at() with more supported method

at() is not supported on some environments, including Hermes, which we rely on for Android.

SiobhyB committed:
refactor: Enable use of debugFactory on native

SiobhyB commented on projects/packages/videopress/src/client/block-editor/hooks/use-video-data/index.ts:
```diff
@@ -72,7 +77,8 @@ export default function useVideoData( {
 
 				// pick filename from the source_url
-				const filename = response.original?.split( '/' )?.at( -1 );
+				const filename =
```
Debugging the reasons this code [wasn't working with Android](https://github.com/Automattic/jetpack/pull/29997#pullrequestreview-1379015953) led me to this line. Although I couldn't find any open issues that specifically mention that Hermes doesn't work with the `at()` method, I did find it isn't supported by [some older browsers](https://caniuse.com/mdn-javascript_builtins_array_at), and replacing it fixes the issue with Android. I therefore strongly suspect the cause of the issues was Hermes not supporting the `at()` method.

Another way of rewriting this could be: `const filename = response.original?.split('/').slice(-1)[0];` but I thought that the use of `length` may be more easily parsed, although it's more verbose. I'd be interested to hear other thoughts or opinions on this.

fluiddot replied:
> Debugging the reasons this code https://github.com/Automattic/jetpack/pull/29997#pullrequestreview-1379015953 led me to this line. Although I couldn't find any open issues that specifically mention that Hermes doesn't work with the at() method, I did find it isn't supported by [some older browsers](https://caniuse.com/mdn-javascript_builtins_array_at), and replacing it fixes the issue with Android. I therefore strongly suspect the cause of the issues was Hermes not supporting the at() method.

Great investigation! It's true that there's no reference to the `at` function in [the Hermes language features](https://hermesengine.dev/docs/language-features/), so as you mentioned might be an incompatibility with the WebView Android version ([it's supported from `92` onwards](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at#browser_compatibility)). However, I checked my Android WebView version and it's `111`, so not sure where's the incompatibility coming from 🤔.

>Another way of rewriting this could be: const filename = response.original?.split('/').slice(-1)[0]; but I thought that the use of length may be more easily parsed, although it's more verbose. I'd be interested to hear other thoughts or opinions on this.

Another approach would be to import a polyfill for the function. I checked the [`at` documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at#see_also) and there's a [polyfill in the `core-js` library that we could use](https://github.com/zloirock/core-js#relative-indexing-method). Technically, we could import it into Gutenberg by adding the following code to [`globals.js `](https://github.com/WordPress/gutenberg/blob/73fb8a63c59b98208c599405ae3071d9b88417b9/packages/react-native-editor/src/globals.js#L10-L13):
```
import 'core-js/proposals/relative-indexing-method';
```

WDYT?

fluiddot replied:
I've just tested my above suggestion on an actual Android device and works 🎊.

fluiddot replied:
@SiobhyB I took the liberty of [opening a PR](https://github.com/WordPress/gutenberg/pull/49749) to include the polyfill in Gutenberg. Let me know if you could take a look and test this PR with the polyfill. Thanks!

SiobhyB replied:
Oh, brilliant, thanks for that! Approved the Gutenberg PR and reverted my change to the code in https://github.com/Automattic/jetpack/pull/29997/commits/bfc9f37215428521c8d539cb357ef88d8bd289bd. I'll also go ahead to update the Gutenberg submodule in https://github.com/wordpress-mobile/gutenberg-mobile/pull/5652 once it's merged.

fluiddot replied:
> Oh, brilliant, thanks for that! Approved the Gutenberg PR and reverted my change to the code in [bfc9f37](https://github.com/Automattic/jetpack/commit/bfc9f37215428521c8d539cb357ef88d8bd289bd). I'll also go ahead to update the Gutenberg submodule in [wordpress-mobile/gutenberg-mobile#5652](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5652) once it's merged.

Great, I've just merged [the PR](https://github.com/WordPress/gutenberg/pull/49749).

SiobhyB replied:
Thank you! The Gutenberg Mobile PR is up-to-date now :) It'd be great if we could get this PR merged now, if you think it's ready to be approved.

SiobhyB commented:
@fluiddot, thank you so much for spotting that! After some digging, I believe the issue lies with Hermes not supporting the `at()` method. I've added my thoughts around the refactor I've done [here](https://github.com/Automattic/jetpack/pull/29997#discussion_r1163867762), and am looking forward to hearing your thoughts. 🙇‍♀️  This PR's ready for review again.

SiobhyB committed:
[not verified] Revert "refactor: Replace at() with more supported method"

This reverts commit 9a3ff4f6f4dccabcfba262ffecf5636fcaf27707.

fluiddot reviewed (approved):
LGTM 🎊 !

I confirm that changes made to the block on the web are reflected when opening the block in the app. I also checked that the block settings are updated even when the post is not refreshed in the app with the latest changes, e.g.:
1. Change block settings on the web.
2. Save the post.
3. Open the same post in the app without refreshing the post list.
4. Observe a log in the console notifying about some of the settings being out of sync (e.g. `'rating' is out of sync. Updating 'rating' attr from 'G' to 'PG-13'`).
